File: python/infinicore/nn/functional/relu6.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def relu6(input: Tensor, inplace: bool = False, *, out: Tensor | None = None) -> Tensor:
    if inplace:
        _infinicore.relu6_(input._underlying, input._underlying)
        return input

    if out is None:
        return Tensor(_infinicore.relu6(input._underlying))

    _infinicore.relu6_(out._underlying, input._underlying)
    return out

File: ops/src/infinicore/ops/relu6/relu6.cc
#include "infinicore/ops/relu6.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Relu6);

Relu6::Relu6(Tensor out, const Tensor &input) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(out, input);
    INFINICORE_GRAPH_OP_DISPATCH(out->device().getType(), out, input);
}

void Relu6::execute(Tensor out, const Tensor &input) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Relu6, out, input);
}

Tensor relu6(const Tensor &input) {
    auto out = Tensor::empty(input->shape(), input->dtype(), input->device());
    relu6_(out, input);
    return out;
}

void relu6_(Tensor out, const Tensor &input) {
    Relu6::execute(out, input);
}

} // namespace infinicore::op
